### PR TITLE
16 add node class

### PR DESCRIPTION
**Pull request by StephenYu-R**
This PR is to add the class Node with a simple class Scene. As building blocks, nodes can be arranged in a scene tree.

This PR commits the following features:
- GetName
- SetName
- AddChild
- AddSibling
- RemoveChild
- GetChildCount
- GetChild
- HasNode
- GetNode
- GetNodeOrNull
- Reparent
- GetTree
- IsInsideTree
- IsAncestorOf
- GetPath
- GetPathTo
- FindCommonParentWith
- MoveChild
- GetIterator
- GetIndex
- GetChildByName
- SetTree
- ValidateChildName
- ValidateNodeName
- Notifications
- Propagation
- GenerateSerialChildName

The following features are not included in this PR:
- Processing
- RPC calls
- Ownership
- Groups
- Complex modes

### CODE REVIEW

**Review by Wuqiqi123 — CHANGES_REQUESTED** [review]

File: src/gobot/scene/node.cpp [hunk 1]
```diff
@@ -3,18 +3,676 @@
...
+    if (tree_changed_b) {
+        Q_EMIT tree_changed_b->treeChanged();
+    }
+}
+
+bool Node::HasNode(const NodePath &path) const {
+    return GetNodeOrNull(path) != nullptr;
+}
+
+void Node::Reparent(Node *parent, bool keep_global_transform) {
```
Comment (Wuqiqi123) [inline src/gobot/scene/node.cpp:393]: Do we need keep_global_transform ?

**Review by Wuqiqi123 — CHANGES_REQUESTED** [review]

File: src/gobot/scene/node.cpp (hunk 1 above)
Comment (StephenYu-R) [reply src/gobot/scene/node.cpp:393]: keep_global_transform will be used to autoset the global transform of a 3D node when it's reparented as this method will be overloaded in Node3D.

Without this option, the global transform should be regulated manually.

File: src/gobot/scene/node.cpp [hunk 2]
```diff
@@ -3,18 +3,676 @@
...
+} // End of namespace gobot
+
+GOBOT_REGISTRATION {
+
+    Class_<Node>("Node")
+            .constructor()(CtorAsRawPtr)
+
+            .property_readonly("child_count", &Node::GetChildCount)
+            .property_readonly("inside_tree", &Node::IsInsideTree)
+            .property_readonly("iterator", &Node::GetIterator)
```
[the remark this answers is not in the document]
Comment (StephenYu-R) [reply src/gobot/scene/node.cpp:658]: removed "iterator"

File: src/gobot/scene/node.cpp (hunk 1 above)
Comment (StephenYu-R) [reply src/gobot/scene/node.cpp:393]: removed this option as Reparent always autoset global transform.

File: include/gobot/error_macros.hpp [hunk 3]
```diff
@@ -55,14 +53,49 @@
...
+
+#define ERR_FAIL_NULL_V_MSG(ptr, ret, msg)                                                      \
+    if (ptr == nullptr) [[unlikely]] {                                                          \
+        LOG_ERROR("Pointer {} is null. {}", GOB_STRINGIFY(ptr), GOB_STRINGIFY(msg));            \
+        return ret;                                                                             \
+    } else                                                                                      \
+        ((void)0)
+
+#define ERR_FAIL_V_MSG(ret, msg)                                                                \
+    if (true) {                                                                                 \
```
[the remark this answers is not in the document]
Comment (StephenYu-R) [reply include/gobot/error_macros.hpp]: removed if statement

**Review by Wuqiqi123 — APPROVED** [review]

**Pull request merged**